interfaces/builtin/dbus_app.go

```diff
@@ -0,0 +1,280 @@
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/release"
+)
+
+// Split this out since we only need these rules once per app
+var dbusAppPermanentSlotAppArmorShared = []byte(`
+# Description: Allow owning a name on DBus public bus
+
+#include <abstractions/###DBUS_BIND_ABSTRACTION###>
+
+# register on DBus
+dbus (send)
+    bus=system
+    path=/org/freedesktop/DBus
+    interface=org.freedesktop.DBus
+    member="{Request,Release}Name"
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+
+dbus (send)
+    bus=system
+    path=/org/freedesktop/DBus
+    interface=org.freedesktop.DBus
+    member="GetConnectionUnix{ProcessID,User}"
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+`)
+
+// These rules are needed for each well-known name for the app
+var dbusAppPermanentSlotAppArmorIndividual = []byte(`
+# bind to a well-known DBus name: ###DBUS_BIND_NAME###
+dbus (bind)
+    bus=###DBUS_BIND_BUS###
+    name=###DBUS_BIND_NAME###,
+
+# Allow us to talk to dbus-daemon
+dbus (receive)
+    bus=###DBUS_BIND_BUS###
+    path=###DBUS_BIND_PATH###
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+dbus (send)
+    bus=###DBUS_BIND_BUS###
+    path=###DBUS_BIND_PATH###
+    interface=org.freedesktop.DBus.Properties
+    peer=(name=org.freedesktop.DBus, label=unconfined),
+`)
+
+var dbusAppPermanentSlotAppArmorIndividualClassic = []byte(`
+# allow unconfined clients talk to ###DBUS_BIND_NAME### on classic
+dbus (receive)
+    bus=###DBUS_BIND_BUS###
+    path=###DBUS_BIND_PATH###
+    peer=(label=unconfined),
+`)
+
+var dbusAppPermanentSlotSecComp = []byte(`
+# Description: Allow owning a name on DBus public bus
+
+getsockname
+recvmsg
+sendmsg
+sendto
+`)
+
+type DbusAppInterface struct{}
+
+func (iface *DbusAppInterface) Name() string {
+	return "dbus-app"
+}
+
+func (iface *DbusAppInterface) PermanentPlugSnippet(plug *interfaces.Plug, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityDBus, interfaces.SecurityAppArmor, interfaces.SecuritySecComp, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *DbusAppInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityDBus, interfaces.SecurityAppArmor, interfaces.SecuritySecComp, interfaces.SecurityUDev, interfaces.SecurityMount:
+		return nil, nil
+	default:
+		return nil, interfaces.ErrUnknownSecurity
+	}
+}
+
+func (iface *DbusAppInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	dbusAppBusNames, err := iface.getBusNames(slot.Attrs)
+	if err != nil {
+		return nil, err
+	}
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		var snippets bytes.Buffer
```

Comment:
(nitpick) You could write `snippets := bytes.NewBufferString("")` here to save 1 line.